Queue.cpp

```diff
@@ -65,29 +74,50 @@ class Queue::Impl
         }
     }
 
-    std::unique_ptr<Msg> request(Msg&& msg)
+    std::unique_ptr<Msg> request(Msg&& msg, int timeoutMillis)
     {
-        // Construct an ad hoc Queue to handle response Msg
+        Request req;
+
+        // emplace the request in the map
         std::unique_lock<std::mutex> lock(responseMapMutex_);
         auto it = responseMap_.emplace(
-            std::make_pair(msg.getUniqueId(), std::unique_ptr<Queue>(new Queue))).first;
-        lock.unlock();
+            std::make_pair(msg.getUniqueId(), &req)).first;
 
         put(std::move(msg));
-        auto response = it->second->get(); // Block until response is put to the response Queue
 
-        lock.lock();
-        responseMap_.erase(it); // Delete the response Queue
-        lock.unlock();
+        // wait_for returns false if the return is due to timeout
+        if (timeoutMillis <= 0)
+            req.condVar.wait(lock, [&req]{return req.response.get();});
+        else
+        {
+            auto timeoutOccured = !req.condVar.wait_for(
+            lock,
+            std::chrono::milliseconds(timeoutMillis),
+            [&req] {return req.response.get();});
+
+            if (timeoutOccured)
+            {
+                // remove the request from the map, as it is on the stack
+                responseMap_.erase(it);
+                return nullptr;
+            }
+        }
+
+        auto response = std::move(it->second->response);
+        responseMap_.erase(it);
 
         return response;
     }
 
     void respondTo(MsgUID reqUid, Msg&& responseMsg)
     {
         std::lock_guard<std::mutex> lock(responseMapMutex_);
-        if (responseMap_.count(reqUid) > 0)
-            responseMap_[reqUid]->put(std::move(responseMsg));
+        auto it = responseMap_.find(reqUid);
+        if(it == responseMap_.end())
+            return;
+
+        it->second->response = responseMsg.move();
+        it->second->condVar.notify_one();
```

Comment:
The lock doesn't need to be held anymore when notifying

Reply:
Good catch!

Reply:
Damn... I think my proposal here was not correct after all...

Think about this scenario:

- Requester thread is waiting in wait()/wait_for() in request()
- Responder thread holds the lock here and is accessing the map
- Requester thread spuriously wakes up/times out -> it starts waiting to get the lock
- Responder thread is done accessing the map here ad unlocks
- Requester thread gets the lock, continues, gets the response from the map and erases the object from the map
- Responder thread tries to notify here -> use after free!

I think our use here falls under this explanation (https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one):

> Notifying while under the lock may nevertheless be necessary when precise scheduling of events is required, e.g. if the waiting thread would exit the program if the condition is satisfied, causing destruction of the notifying thread's condition_variable. A spurious wakeup after mutex unlock but before notify would result in notify called on a destroyed object.

Sorry about my wrong suggestion and unnecessary work! These threading things can be tricky sometimes...

Reply:
Maybe a comment could be added that it's necessary to hold the lock while notifying here